### PR TITLE
feat: Add options to disable case insensitivity and force email and username to lower case

### LINKED ISSUE #7937: Parser Server not fully compatible with Mongo Atlas Serverless
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
I currently checking if parse-server could run in a fully serverless env.
The experiment is based on GCP Cloud Run and the new Mongo Atlas Serverless service (https://www.mongodb.com/cloud/atlas/serverless)

I found an issue (no current workarounds) about collation (case insensitive login/query/indexes). Mongo Serverless does not support collation. Developers are not able to "disable" collation. The collation fails will for example block logins, and block Defined Schema migration.

### Steps to reproduce
Run a parse-server and connect to a Mongo Serverless Instance

### Actual Outcome
All stuff related to case insensitive query/index will fail

### Expected Outcome
To be able to disable the collation (mainly used on the Parse. User class)
### Environment
[5.3.0-alpha.6](https://www.npmjs.com/package/parse-server/v/5.3.0-alpha.6)

Server
- Parse Server version: `5.3.0-alpha.6`

Database
- System (MongoDB or Postgres): `Mongo Atlas Serverless`
- Database version: `5.3.1`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `remote`

### Logs
Collation is not supported on this Altas tier (Mongo Serverless)

### PR DESCRIPTION
**Pull request by Moumouls**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
Add options to be able to support MongoDB Serverless (no collation support)

Related issue: #7937 

### Approach
Add 2 abstract options disableCaseInsensitivity and forceEmailAndUsernameToLowerCase to allow a developer to disable default collations of Parse Server and forceEmailAndUsernameToLowerCase to keep clean usernames and emails in the database.

### TODOs before merging

- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

**Comment by mtrezza** [comment]
Can we find something that reads easier than "disable case insensitivity" - avoiding the double negation. It could also be more descriptive - "case sensitivity for what"?

File: src/Controllers/DatabaseController.js [hunk 1]
```diff
@@ -360,6 +360,15 @@ const relationSchema = {
   fields: { relatedId: { type: 'String' }, owningId: { type: 'String' } },
 };
 
+const maybeTransformUsernameAndEmailToLowerCase = (object, className, options) => {
```
Comment (mtrezza) [inline src/Controllers/DatabaseController.js:368]: - remove `maybe`
- use one term across all code, currently we have "force" and "transform", let's use "convert" for all

File: src/Options/index.js [hunk 2]
```diff
@@ -94,6 +94,12 @@ export interface ParseServerOptions {
   databaseOptions: ?DatabaseOptions;
   /* Adapter module for the database */
   databaseAdapter: ?Adapter<StorageAdapter>;
+  /* Disable case insensitivity (collation) on queries and indexes, needed if the you use mongodb serverless
+  :ENV: PARSE_SERVER_DISABLE_DEFAULT_CASE_INSENSITIVITY */
```
Comment (mtrezza) [inline src/Options/index.js:106]: - Is it necessary to specify the ENV parameter? The options script should create that automatically from the param name. That also prevents the mistake, that the env var name is different than the param name, as is here.
- Please use correct spelling ("MongoDB") and notation, this should be spelled like full sentences, as it goes into the API docs as prose text. 
- Can we find a param name that is easier to read and more descriptive, for example avoiding the double negation, for example `enable...` with a default value of `true`?

**Comment by Moumouls** [comment]
Currently, case-insensitivity is only used for indexes/queries on email/username fields on User class.

I can suggest:
- preventCollation
- disableCaseInsensitiveIndexAndQuery
- preventCollationOnIndexAndQuery

Collation is also the error name returned by the Mongo driver if you try to use "case insensitive" index/query on Mongo Serverless. Using the same wording could help developers to fix the error.

I do not have other ideas @mtrezza  🙂

File: src/Options/index.js [hunk 3]
```diff
@@ -94,6 +94,12 @@ export interface ParseServerOptions {
...
+  disableCaseInsensitivity: ?boolean;
+  /* Force username and email to lowercase on create/update/login/signup. On queries client needs to ensure to send lowercase email/username
+  :ENV: PARSE_SERVER_FORCE_EMAIL_AND_USERNAME_TO_LOWER_CASE */
+  forceEmailAndUsernameToLowerCase: ?boolean;
```
Comment (mtrezza) [inline src/Options/index.js:110]: - Instead of "force" let call it "convert to".
- Notation as if these were full sentences, this goes into the API docs.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Is the collation applied for all actions regarding username / email, e.g. also password reset, email verification?

**Comment by mtrezza** [comment]
> Collation is also the error name returned by the Mongo driver 

Maybe lets call it `enableCollation...`. It may be useful to split up the collation for email / username into 2 separate params to make this more useful, also outside of a server-less usage. Email collation is also somewhat sensitive, because by specification `email@example.com` is not the same as `EMAIL@example.com`, these are technically 2 different mailboxes.

**Comment by Moumouls** [comment]
Yes `enableCollation`, `transformEmailToLowerCase` and `transformUsernameToLowerCase` makes sense.

Do you have a suggestion about "default injection" of `enableCollation` to `true`. Or i just use a `=== false` check in the code. It means that if `enableCollation` does not exists in the config, it's enabled @mtrezza

File: src/Controllers/DatabaseController.js (hunk 1 above)
Comment (Moumouls) [reply src/Controllers/DatabaseController.js:368]: we can remove `maybe` but i'll keep transform. It's the term used in this file


```js
              update = transformObjectACL(update);
              transformUsernameAndEmailToLowerCase(update, className, this.options);
              transformAuthData(className, update, schema);
```

File: src/Options/index.js (hunk 3 above)
Comment (Moumouls) [reply src/Options/index.js:110]: what about ? `transformUsernameToLowerCase` and `transformEmailToLowerCase`

**Comment by mtrezza** [comment]
Let's simply use `convert`, it's a simple case conversion.

I don't understand the 2nd part.

File: src/Controllers/DatabaseController.js [hunk 4]
```diff
@@ -360,6 +360,15 @@ const relationSchema = {
...
+  if (className === '_User' && options.forceEmailAndUsernameToLowerCase) {
+    const toLowerCaseFields = ['email', 'username'];
+    toLowerCaseFields.forEach(key => {
+      if (typeof object[key] === 'string') object[key] = object[key].toLowerCase();
```
Comment (dblythy) [inline src/Controllers/DatabaseController.js]:
```suggestion
      if (typeof object[key] === 'string') {
        object[key] = object[key].toLowerCase();
      }
```

File: src/Options/Definitions.js [hunk 5]
```diff
@@ -161,6 +161,12 @@ module.exports.ParseServerOptions = {
     action: parsers.booleanParser,
     default: false,
   },
+  disableCaseInsensitivity: {
+    env: 'PARSE_SERVER_DISABLE_DEFAULT_CASE_INSENSITIVITY',
+    help:
+      'Disable case insensitivity (collation) on queries and indexes, needed if the you use mongodb serverless',
```
Comment (dblythy) [inline src/Options/Definitions.js]:
```suggestion
      'Disable case insensitivity (collation) on queries and indexes, required for serverless deployments',
```

File: src/Options/Definitions.js [hunk 6]
```diff
@@ -227,6 +233,12 @@ module.exports.ParseServerOptions = {
     action: parsers.objectParser,
     default: {},
   },
+  forceEmailAndUsernameToLowerCase: {
+    env: 'PARSE_SERVER_FORCE_EMAIL_AND_USERNAME_TO_LOWER_CASE',
+    help:
+      'Force username and email to lowercase on create/update/login/signup. On queries client needs to ensure to send lowercase email/username',
```
Comment (dblythy) [inline src/Options/Definitions.js:239]: Not really sure what this means

**Review by dblythy — CHANGES_REQUESTED** [review]
Some nice refactoring! Minor suggestions below.

File: src/Controllers/DatabaseController.js (hunk 1 above)
Comment (mtrezza) [reply src/Controllers/DatabaseController.js:363]: Which file? A "transformation" implies a more complex change of structure, this is simply a case conversion. Please use convert.

File: src/Options/index.js (hunk 3 above)
Comment (mtrezza) [reply src/Options/index.js:102]: See https://github.com/parse-community/parse-server/pull/8042#discussion_r912342663

File: src/Controllers/DatabaseController.js (hunk 4 above)
Comment (mtrezza) [reply src/Controllers/DatabaseController.js]: Is there a lint rule for this? This pattern should really be prevented. https://github.com/parse-community/parse-server/issues/8093

**Comment by Moumouls** [comment]
Real usage feedback: With this PR and options activated I currently succeed in running Parse Server in a fully Serverless env using Google Cloud Run (CPU only reserved during requests) and MongoDB Atlas Serverless and it works pretty well.

**Comment by Moumouls** [comment]
okay @mtrezza  @dblythy I'll refresh this one with your feedback

**Comment by mtrezza** [comment]
To go ahead and merge this, could you address any open comments?

**Comment by lilonpro** [comment]
Does this resovle issue #8492 ? I just tried to run this PR and still hitting the same error message.

**Comment by mattia1208** [comment]
Real usage feedback: 
With the PR of @Moumouls and options activated I currently succeed in running Parse Server in a fully Serverless env using Azure Web App + Azure Cosmos DB for MongoDB and it works pretty well.

I think by now it is necessary to make Parse Server compatible with serverless environments.

**Comment by mtrezza** [comment]
@mattia1208 How is your post related to this issue?

**Comment by mtrezza** [comment]
@mattia1208 As you can see in the conversation this PR has open issues that need to be resolved to get merged. Maybe @Moumouls could review them to see which ones can be closed. And there is a merge conflict to resolve.

**Comment by Moumouls** [comment]
Hi @mattia1208 , happy to here that adjustments that i made allow to run parse server with Azure cosmos DB.

It's because Azure Cosmos like Mongo Serverless doesn't support insensitive indexes ? 


A special not @mattia1208 , Parse Server is quite slow to start form a cold start, what is your serverless setup on Azure to prevent cold start. On my side on GCP Cloud Run i've some minute cron task needed for special events, it helps to keep atleast one instance active wihtout paying for a full hot docker instance.

**Comment by mattia1208** [comment]
Hi @Moumouls ,

Yes, you're correct. Azure Cosmos, like Mongo Serverless, does not support case insensitive indexes.

For my serverless setup on Azure, I use the following:

Azure Web App + Database + Redis Cache
The Web App is hosted on Linux using a Premium V2 SKU, specifically a small size with a total of 210 ACU and 3.5GB of     memory.
I also utilize Azure Cosmos DB with the API for MongoDB.

With this setup, I haven't experienced any cold start issues.

I hope this provides clarity!

**Comment by b10f** [comment]
@Moumouls, @mtrezza,

I recently came across this PR while searching for a solution for a particular problem. 

Long story short, AWS’ DocumentDB also suffers from the same issue that this PR is intended to solve. 

To have this PR merged would be a GIGANTIC help for our team, if we could utilize AWS DocumentDB Elastic Clusters to scale our database horizontally. I tried to install @Moumouls version from this exact branch from this PR and it works absolutely beautifully.

Is there any chance that you gentlemen can make this happen in the foreseeable future? Is there anything I can help with to make this happen? 

Thanks.

**Comment by mtrezza** [comment]
@b10f see https://github.com/parse-community/parse-server/pull/8042#issuecomment-1684308292; if you want to help, you could either suggest the changes via PR review (use code suggestion), or use this PR as a start for a new PR and incorporate the changes.

File: src/Options/index.js (hunk 3 above)
Comment (mtrezza) [reply src/Options/index.js:110]: Sounds good

**Comment by b10f** [comment]
@Moumouls, any thoughts?

**Comment by Moumouls** [comment]
Hi @mtrezza @b10f @mattia1208 @lilonpro , if someone want to recycle my code and finish last details feel free to continue this one.

My team is moving out of Mongo Serverless for some specific reason ( but Mongo Serverless works really well with this PR), if it helps your teams now feel free to finish last details of this PR 🚀 

We are close, but i will dedicate some time to parse-server on other PRs like GraphQL.

**Comment by b10f** [comment]
@Moumouls, since you are the original author of this PR, and considering that it only needs some merge conflicts to be sorted out, and addressing some review comments, I'm assuming it would not take too much effort.

Any chance that you could finish it? Even though you're moving out from Mongo Serverless, this PR would make it possible to use AWS DocumentDB and other providers that still have the same limitation.

**Comment by Moumouls** [comment]
> It only needs some merge conflicts to be sorted out and addressing some review comments.

Actually, it's not difficult to finish the PR. I'm a volunteer contributor to Parse Server, and most of the time, I try to send PRs to parse-server when I discover an improvement. However, there is no magic, unfortunately. I work and dedicate my time to parse-server on features that are really used by my team. Now we are moving away from Mongo Serverless, and we do not use DocumentDB, so we are not "motivated" to ship this ASAP 😉. I keep my open-source time for other PRs/projects.

Here, the job is almost done. Since we are in an open-source environment, you can easily fork, copy the branch, and provide suggested changes by @mtrezza, and it's done. If it's your first contribution to parse-server, it might take some additional time, but contributing to parse-server is feasible. Even better, you will be able to easily contribute in the future 🚀.

I'll be happy to help you start contributing 💯. @b10f

**Comment by mtrezza** [comment]
I have also simplified the comments and removed some of the less important suggested changes (like naming), so it will hopefully be easier to get this PR finished.

**Comment by b10f** [comment]
@mtrezza https://github.com/parse-community/parse-server/pull/8767

**Comment by mtrezza** [comment]
Continued in https://github.com/parse-community/parse-server/pull/8805

**Pull request closed without merging**